pkg/pipeline/conntrack/conntrack_test.go

```diff
@@ -166,10 +168,187 @@ func TestTrack(t *testing.T) {
+			"20s: flow BA",
+			startTime.Add(20 * time.Second),
+			[]config.GenericMap{flBA4},
+			[]config.GenericMap{
+				flBA4,
+			},
+		},
+		{
+			"49s: no end conn",
+			startTime.Add(49 * time.Second),
+			nil,
+			nil,
+		},
+		{
+			"51s: end conn BA",
+			startTime.Add(51 * time.Second),
+			nil,
+			[]config.GenericMap{
+				newMockRecordConnAB(ipA, portA, ipB, portB, protocol, 333, 777, 33, 77, 4),
+			},
+		},
+	}
+
+	for _, test := range table {
+		var prevTime time.Time
+		t.Run(test.name, func(t *testing.T) {
+			require.Less(t, prevTime, test.time)
+			prevTime = test.time
+			clk.Set(test.time)
+			actual := ct.Track(test.inputFlowLogs)
+			require.Equal(t, test.expected, actual)
+		})
+	}
+}
+
+// TestEndConn_Unidirectional tests that end connection records are outputted correctly and in the right time in
+// unidirectional setting.
+// The test simulates 2 flow logs from A to B and 2 from B to A in different timestamps.
+// Then the test verifies that an end connection record is outputted only after 30 seconds from the last flow log.
+func TestEndConn_Unidirectional(t *testing.T) {
+	clk := clock.NewMock()
+	conf := buildMockConnTrackConfig(false, []string{"newConnection", "flowLog", "endConnection"})
+	ct, err := NewConnectionTrack(*conf, clk)
+	require.NoError(t, err)
+
+	ipA := "10.0.0.1"
+	ipB := "10.0.0.2"
+	portA := 9001
+	portB := 9002
+	protocol := 6
+
+	flAB1 := newMockFlowLog(ipA, portA, ipB, portB, protocol, 111, 11)
+	flAB2 := newMockFlowLog(ipA, portA, ipB, portB, protocol, 222, 22)
+	flBA3 := newMockFlowLog(ipB, portB, ipA, portA, protocol, 333, 33)
+	flBA4 := newMockFlowLog(ipB, portB, ipA, portA, protocol, 444, 44)
+	startTime := clk.Now()
+	table := []struct {
+		name          string
+		time          time.Time
+		inputFlowLogs []config.GenericMap
+		expected      []config.GenericMap
+	}{
+		{
+			"start: flow AB",
+			startTime.Add(0 * time.Second),
+			[]config.GenericMap{flAB1},
+			[]config.GenericMap{
+				newMockRecordConn(ipA, portA, ipB, portB, protocol, 111, 11, 1),
+				flAB1,
+			},
+		},
+		{
+			"10s: flow AB and BA",
+			startTime.Add(10 * time.Second),
+			[]config.GenericMap{flAB2, flBA3},
+			[]config.GenericMap{
+				flAB2,
+				newMockRecordConn(ipB, portB, ipA, portA, protocol, 333, 33, 1),
+				flBA3,
+			},
+		},
+		{
+			"20s: flow BA",
+			startTime.Add(20 * time.Second),
+			[]config.GenericMap{flBA4},
+			[]config.GenericMap{
+				flBA4,
+			},
+		},
+		{
+			"39s: no end conn",
+			startTime.Add(39 * time.Second),
+			nil,
+			nil,
+		},
+		{
+			"41s: end conn AB",
+			startTime.Add(41 * time.Second),
+			nil,
+			[]config.GenericMap{
+				newMockRecordConn(ipA, portA, ipB, portB, protocol, 333, 33, 2),
+			},
+		},
+		{
+			"49s: no end conn",
+			startTime.Add(49 * time.Second),
+			nil,
+			nil,
+		},
+		{
+			"51s: end conn BA",
+			startTime.Add(51 * time.Second),
+			nil,
+			[]config.GenericMap{
+				newMockRecordConn(ipB, portB, ipA, portA, protocol, 777, 77, 2),
+			},
+		},
```

Comment:
What identifying mark is there for connection entries that are completed? Where do you test for that?

Reply:
Good points. There isn't a mark yet. I'm working on adding a field to indicate the record type and submit it in a separate PR.